### ISSUE TITLE
[issue]privatewin10 firewall rules window & high cpu usage, freeze

### ISSUE THREAD

**Issue by whismerhill** [DavidXanatos/priv10#63]
hi,
so I'm not sure exactly what are the triggers for this :
-I'm in the firewall rules window doing stuff, the exact thing I'm doing can be as simple as scrolling around
and then privatewin10 window freezes & stop responding
-task manager shows about 50% cpu usage, process explorer shows 33%
(that's with an i7-2600k)
-and then I have to kill the program

if you want me to do anything just tell me

**Comment by whismerhill** [comment]
update : 
-fresh reboot
-open program
-firewall tab sort by description
-scroll a bit
-freeze

**Comment by DavidXanatos** [comment]
what view do you have on the program list the compact or the verbose one?

**Comment by Mrnofish** [comment]
Priv10 is unusable for me since around the 2004 upgrade, so I have temporarily switched to a different product.

I miss some of Priv10's features, however the combination of extreme slowness and constant crashes made it essentially impossible to manage existing rules and create new ones.

Flags about this situation have been already signaled a long time ago, so the product appears all but abandoned at this time.

I'm still subscribed to releases and I'll be happy to consider Priv10 again if the performance returns to acceptable.

**Comment by DavidXanatos** [comment]
The tool is not abandoned don't wary I'm just very busy with sandboxie in the last months.
A new build with major reworks is already in the making but it will take a few more weeks to complete.

**Comment by Mrnofish** [comment]
> The tool is not abandoned don't wary I'm just very busy with sandboxie in the last months.

Your singular commitment to this and other projects is understood and appreciated, but for a tool like a firewall manager, it is hardly acceptable for it to be left in a state this broken for so long. Even with most rules in place, a firewall manager is not an optional component I can just squint and pretend the issues are not there. I can only imagine how fast a new user would run away.

No hard feelings, though, just trying to convey what this looks like from one user's perspective.

**Comment by whismerhill** [comment]
> what view do you have on the program list the compact or the verbose one?

I have no idea which one is this : https://prnt.sc/uoqgl5
![download](https://user-images.githubusercontent.com/17818248/94365423-30185d00-00d1-11eb-9e89-f13a2d770a35.png)
if needed I can make a video showing what happens

**Comment by whismerhill** [comment]
> I miss some of Priv10's features, however the combination of extreme slowness and constant crashes made it essentially impossible to manage existing rules and create new ones.

no slowness here, just freezes, but maybe if you have a better performing computer, my "freezes" are just slowness for you...
that said, I have been able to use the product anyway, it's just a bit of a bother sometime

**Comment by DavidXanatos** [comment]
try the new 0.8x line of builds pelase

**Comment by whismerhill** [comment]
> try the new 0.8x line of builds pelase

I don't see much change I'm afraid
I could tell a bunch of stuff but that would be suppositions on my part so ....

**Comment by DavidXanatos** [comment]
One part I know is working not great is the sorting of the verbose program set list.
Improving re doing that is on my todo list but for some time later.

From your screenshot I see you are using that view.
Switching to the compact view for testing would be helpful to confirm that its this particular tree view making the troubles.

The compact view:
![grafik](https://user-images.githubusercontent.com/3890945/96338535-0b3f5600-108f-11eb-934a-2325722b84c9.png)

**Comment by whismerhill** [comment]
ho wow ! problem "fixed" (or rather worked around) by changing to that view indeed you were right

Thanks a bunch, that will alleviate my issue until you fix the other views.

**Comment by DavidXanatos** [comment]
please try if this build solves the issues with the verbose view

**Comment by whismerhill** [comment]
0.83 you mean ?
with 0.83,  it's even worse
somehow it changed automatically when going from settings back to firewall the compact view was gone
and the verbose view took its place
I only moused over the view that's all (didn't click anywhere)

and now each time I close & startup the program it is immediately frozen on this view and I can't even get out of it
so basically 0.83 is almost unusable, I only managed to use it for at most 5 minutes.

**Comment by DavidXanatos** [comment]
that is strange here it works well...

to ensure you start in the compact view set the following values in the PrivatwWin10.ini

[GUI]
FirewallViewMode=NormalView
FirewallProgsWidth=272

**Comment by DavidXanatos** [comment]
can you start the UI from a command prompt and see if its loging any issue there?

**Comment by DavidXanatos** [comment]
Please try this build, does it improve the situation?

Its difficult to fix something I cant reproduce, I can only guess and try to fix on a hunch

[PrivateWin10.zip](https://github.com/DavidXanatos/priv10/files/5404688/PrivateWin10.zip)

**Comment by whismerhill** [comment]
> that is strange here it works well...
> 
> to ensure you start in the compact view set the following values in the PrivatwWin10.ini
> 
> [GUI]
> FirewallViewMode=NormalView
> FirewallProgsWidth=272

thanks that helped make the program work again


> Please try this build, does it improve the situation?
> 
> Its difficult to fix something I cant reproduce, I can only guess and try to fix on a hunch
> 
> [PrivateWin10.zip](https://github.com/DavidXanatos/priv10/files/5404688/PrivateWin10.zip)

no changes, going to verbose immediately freezes the program

yeah I know, it's difficult. I wish I could help more. Although who knows.... I have a CPU + motherboard + mem upgrade that should be delivered to me tomorrow ........

**Comment by DavidXanatos** [comment]
How many program entries do you have in total in your program list, even with a old PC it shouldn't freeze for ever.
The initial list creation for the first time may take a second or two but that should be only on the start when the list is built for the  first time.

**Comment by whismerhill** [comment]
when going verbose and since the ini edit, it doesn't freeze immediately, only after scrolling a bit

I just left it alone for at least 20 minutes, sitting at ~50% cpu on a 2600k @4.5Ghz
I don't think it'll unfreeze after that

as for how many programs I have as rules, no idea to be honest, quite a few, maybe nearly twice as many as what windows comes with by default ?

**Comment by whismerhill** [comment]
ok so .... how should I say that
this bug is hardware dependent I think
upgraded from i7-2600k + P67 + 16GB 
to Ryzen 5 3600 + X570 + 32 GB
and the problem vanished

I'd hazard it might be a cpu micro code thing ? but I could be wrong
so... I'm tempted to close this; but then it's not really "solved"

**Comment by whismerhill** [comment]
so .... the freezing vanished
but here's what I saw today : 
![Screenshot_9](https://user-images.githubusercontent.com/17818248/96921620-1c350080-14af-11eb-8d5d-a0c3733df620.png)

unsure if I should open a new issue, seems like with my ryzen 5 3600, there's no more freezing, because the cpu can actually handle the load, but seems like sometimes private win10 is eating cpu cycles for breakfast....
the fact that private win10 ate 34 hours of cputime VS idle 49 hours is a pretty big indicator imho (normally idle represents over 80~90% of the cpu time)

**Comment by DavidXanatos** [comment]
And this cpu cycle eating happens only when the verbose view is active?
Does it correlate with sorting by any column?
Do you sort by a particular column preferably?
IMHO once all the items are inserted i.e. after the first list update the sorting can be the only operation that causes a high CPU load.

**Comment by whismerhill** [comment]
I don't know how but if there's a way to send you a private message, maybe I could send to you a short youtube video link showing the increased cpu usage that stays that way afterward for ever...
(reason for private message is : I don't wanna put up a video in public of my stuff ...)

> And this cpu cycle eating happens only when the verbose view is active?

no it happens as soon as the verbose view comes into view & stays afterward until private win10 is closed (yes even hours later)
> Does it correlate with sorting by any column?
> Do you sort by a particular column preferably?

sorted by name didn't really play with sorting
> IMHO once all the items are inserted i.e. after the first list update the sorting can be the only operation that causes a high CPU load.

yah populating sometimes takes a little bit of time indeed, afterward it's fine, but if the verbose view is shown then the cpu cycles are eaten away, and this "cpu cycle eating behavior" stays until private win10 is closed 
in other words neither : minimizing, returning the view to compact, going back to program settings  (instead of firewall view display)
nothing of the above has any impact.

Also just noticed when closing the window after it eats cpu cycles I get two errors in windows event viewer :
`Faulting application name: PrivateWin10.exe, version: 0.83.0.0, time stamp: 0xd5b60497
Faulting module name: KERNELBASE.dll, version: 10.0.18362.1049, time stamp: 0x1b835a75
Exception code: 0xe0434352
Fault offset: 0x0000000000043b29
Faulting process id: 0x1554
Faulting application start time: 0x01d6a8b8e90a733f
Faulting application path: C:\Program Files\PrivateWin10\PrivateWin10.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: 1fc5a897-409b-4a59-8d7c-6689860ee2da
Faulting package full name: 
Faulting package-relative application ID: `
_____________________________________________________
`Application: PrivateWin10.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Threading.Tasks.TaskCanceledException
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at System.Windows.Threading.DispatcherOperation.Wait(System.TimeSpan)
   at System.Windows.Threading.Dispatcher.InvokeImpl(System.Windows.Threading.DispatcherOperation, System.Threading.CancellationToken, System.TimeSpan)
   at System.Windows.Threading.Dispatcher.Invoke(System.Action, System.Windows.Threading.DispatcherPriority, System.Threading.CancellationToken, System.TimeSpan)
   at System.Windows.Threading.Dispatcher.Invoke(System.Action)
   at PrivateWin10.Controls.TreeItem.<GetIcon>b__12_0(System.Windows.Media.ImageSource)
   at System.Runtime.Remoting.Messaging.AsyncResult.SyncProcessMessage(System.Runtime.Remoting.Messaging.IMessage)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.AsyncProcessMessage(System.Runtime.Remoting.Messaging.IMessage, System.Runtime.Remoting.Messaging.IMessageSink)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
`

**Comment by DavidXanatos** [comment]
you can email me at gmail.com m user name there is xanatosdavid
its really a pitty that github does not have any sort of private messaging system.

**Comment by DavidXanatos** [comment]
What I find extremely strange is that you said the high CPU usage remains even after switching back to the Compact View, as than no updates to the Tree View should be performed hence it should not consume any CPU anymore.

**Comment by xiao-mantou** [comment]
Build 0.84
Freeze for hundreds of milliseconds every few seconds, only in win firewall screen, low cpu usage.
Build v0.75b works fine.
btw afterburner have the same issue for me.

**Comment by DavidXanatos** [comment]
So the CPU usage is low but its freezing from time to time, does this happen now in booth views or only in the verbose one?

whats afterburner ?

EDIT: could you please test more detailed witch which build the issue got introduced.

**Comment by xiao-mantou** [comment]
~Both I think. Priv10 is really buggy now. I changed view few times(final view is the compact view) and the issue temporarily gone, but when I change it again window disappeared. Got `Failed to start engine process` when I try to launch it again.~

Ah wait, after disable Real-time protection 0.84 works fine for now.
12hours ago:
![Screenshot (3393)](https://user-images.githubusercontent.com/25640395/97128592-acd04280-1734-11eb-82be-d6323380fb4f.png)
My problem, can be fixed, BUT today:
![Screenshot (3392)](https://user-images.githubusercontent.com/25640395/97128720-f751bf00-1734-11eb-8a5e-48913a6af104.png)
![Screenshot (3395)](https://user-images.githubusercontent.com/25640395/97128726-fae54600-1734-11eb-89ed-c37cb3f9866c.png)
So [here](https://www.microsoft.com/en-us/wdsi/filesubmission).

[MSI Afterburner](https://www.guru3d.com/files-details/msi-afterburner-beta-download.html) is another story. And he got no luck today.

**Comment by xiao-mantou** [comment]
Hello, this issue still exists.
`Monitor host names queries(DNS Inspector)` might be the real reason.
Disable this function freeze disappeared immediately.
Same in 0.75b. But I am no sure if its the only reason.
In those new versions, I can't manage the firewall at the first launch. 
And I think that would be nice if we will be able to copy hostname from DNS inspector.

**Comment by DavidXanatos** [comment]
> In those new versions, I can't manage the firewall at the first launch.

what do you mean? this small first start wizard window?
you could be able to manage the firewall once you enable it in the settings.
can you make a video of the problem so i see exactly what you mean.

**Comment by xiao-mantou** [comment]
> you could be able to manage the firewall once you enable it in the settings.

Yes, but except the first launch. At **the first launch(with small window pop up)**, firewall is disabled whatever the setting is.
![Snipaste_2020-10-27_20-05-35](https://user-images.githubusercontent.com/25640395/97300793-d3829c00-184e-11eb-8f39-2437385ef857.png)

**Comment by magicoflolis** [comment]
Last night @ 1 in the morning while I was on YouTube I had a similar issue with Priv10 suddenly taking up my entire RAM, CPU, & Disk usage. _I forgot to capture a screenshot_
* I'll include a snip of the logs you see on the homepage & a screenshot of my current settings.
![image](https://user-images.githubusercontent.com/10752602/99121508-5b87d600-25b1-11eb-8539-5bba1f6e3aa2.png)
[logs.txt](https://github.com/DavidXanatos/priv10/files/5539365/logs.txt)

**Comment by g371** [comment]
I have the same issue in every version I have used. Just updated to v0.85, the same. Works just dragging the left column smaller so that it switches to the compact view (took me a while to guess that it's how you can get compact view)

I can reproduce that in full table view by just scrolling a bit, clicking on apps in the list and it freezes so bad, that I can just kill the process. CPU jumps and stays there (Intel Core i5-8250U, 24G ram, Windows 10)

![Clip_3](https://user-images.githubusercontent.com/18262108/134981272-18a490c5-d7e9-44eb-b2af-0e82e479b107.jpg)

- Unchecked Monitor host names queries(DNS Inspector)
- Windows Defender completely disabled

Created C:\Program Files\PrivateWin10\PrivateWin10.ini - seems that it has no effect on view or that works just in portable install? 

Thanks for otherwise great software.
